includes/Core/Util/REST_Entity_Search_Controller.php

```diff
@@ -0,0 +1,148 @@
+<?php
+/**
+ * Class Google\Site_Kit\Core\Util\REST_Entity_Search_Controller
+ *
+ * @package   Google\Site_Kit\Core\Util
+ * @copyright 2021 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\Util;
+
+use Google\Site_Kit\Context;
+use Google\Site_Kit\Core\Permissions\Permissions;
+use Google\Site_Kit\Core\REST_API\REST_Route;
+use WP_REST_Server;
+use WP_REST_Request;
+use WP_REST_Response;
+
+/**
+ * Class for handling entity search rest routes.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+class REST_Entity_Search_Controller {
+
+	/**
+	 * Plugin context.
+	 *
+	 * @since n.e.x.t
+	 * @var Context
+	 */
+	private $context;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Context $context        Plugin context.
+	 */
+	public function __construct( Context $context ) {
+		$this->context = $context;
+	}
+
+	/**
+	 * Registers functionality through WordPress hooks.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function register() {
+		add_filter(
+			'googlesitekit_rest_routes',
+			function ( $routes ) {
+				return array_merge( $routes, $this->get_rest_routes() );
+			}
+		);
+	}
+
+	/**
+	 * Gets REST route instances.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return REST_Route[] List of REST_Route objects.
+	 */
+	protected function get_rest_routes() {
+		$can_authenticate = function() {
+			return current_user_can( Permissions::AUTHENTICATE );
+		};
+
+		return array(
+			new REST_Route(
+				'core/search/data/entity-search',
+				array(
+					array(
+						'methods'             => WP_REST_Server::READABLE,
+						'callback'            => function( WP_REST_Request $request ) {
+							$query = rawurldecode( $request['query'] );
+							$entities = array();
+							if ( filter_var( $query, FILTER_VALIDATE_URL ) ) {
+								$entity = $this->context->get_reference_entity_from_url( $query );
+								if ( $entity && $entity->get_id() ) {
+									$entities = array(
+										array(
+											'id'    => $entity->get_id(),
+											'title' => $entity->get_title(),
+											'url'   => $entity->get_url(),
+											'type'  => $entity->get_type(),
+										),
+									);
+								}
+							} else {
+								$args = array(
+									'posts_per_page'  => 10,
+									'google-site-kit' => 1,
+									's'               => $query,
+									'no_found_rows'   => true,
+									'update_post_meta_cache' => false,
+									'update_post_term_cache' => false,
+									'post_status'     => array( 'publish' ),
+								);
+
+								$posts = ( new \WP_Query( $args ) )->posts;
+
+								if ( ! empty( $posts ) ) {
+									$entities = array_map(
+										function( $post ) {
+											$entity = new Entity(
```

Comment:
Given the IB mentions using [`create_entity_for_post`](https://github.com/google/site-kit-wp/blob/4713411/includes/Core/Util/Entity_Factory.php#L169-L186), it looks like we should change the access modifier for `create_entity_for_post` to `public` and then use it here rather than instantiating `Entity` directly. I.e.

```php
$entity = Entity_Factory::create_entity_for_post( $post );
```

Reply:
Thanks for spotting it, I initially wanted to do it, but if felt drastic to change access modifiers. Let's hope changing it doesn't break anything. :pray: